src/dataflow/src/sink/kafka.rs

```diff
@@ -27,8 +33,73 @@ use repr::{RelationDesc, Row};
 
 use super::util::sink_reschedule;
 
+/// Per-Kafka sink metrics.
+pub struct SinkMetrics {
+    messages_sent_counter: IntCounter,
+    message_send_errors_counter: IntCounter,
+    message_delivery_errors_counter: IntCounter,
+    rows_queued: UIntGauge,
+    messages_in_flight: UIntGauge,
+}
+
+impl SinkMetrics {
+    fn new(topic_name: &str, sink_id: &str, worker_id: &str) -> SinkMetrics {
+        lazy_static! {
+            static ref MESSAGES_SENT_COUNTER: IntCounterVec = register_int_counter_vec!(
+                "mz_kafka_messages_sent_total",
+                "The number of messages the Kafka producer successfully sent for this sink",
+                &["topic", "sink_id", "worker_id"]
+            )
+            .unwrap();
+            static ref MESSAGE_SEND_ERRORS_COUNTER: IntCounterVec = register_int_counter_vec!(
+                "mz_kafka_message_send_errors_total",
+                "The number of times the Kafka producer encountered an error on send",
+                &["topic", "sink_id", "worker_id"]
+            )
+            .unwrap();
+            static ref MESSAGE_DELIVERY_ERRORS_COUNTER: IntCounterVec = register_int_counter_vec!(
+                "mz_kafka_message_delivery_errors_total",
+                "The number of messages that the Kafka producer could not deliver to the topic",
+                &["topic", "sink_id", "worker_id"]
+            )
+            .unwrap();
+            static ref ROWS_QUEUED: UIntGaugeVec = register_uint_gauge_vec!(
+                "mz_kafka_sink_rows_queued",
+                "The current number of rows queued by the Kafka sink operator (note that one row can generate multiple Kafka messages)",
+                &["topic", "sink_id", "worker_id"]
+            )
+            .unwrap();
+            static ref MESSAGES_IN_FLIGHT: UIntGaugeVec = register_uint_gauge_vec!(
+                "mz_kafka_sink_messages_in_flight",
+                "The current number of messages waiting to be delivered by the Kafka producer",
+                &["topic", "sink_id", "worker_id"]
+            )
+            .unwrap();
+        }
+        let labels = &[topic_name, sink_id, worker_id];
+        SinkMetrics {
+            messages_sent_counter: MESSAGES_SENT_COUNTER.with_label_values(labels),
+            message_send_errors_counter: MESSAGE_SEND_ERRORS_COUNTER.with_label_values(labels),
+            message_delivery_errors_counter: MESSAGE_DELIVERY_ERRORS_COUNTER
+                .with_label_values(labels),
+            rows_queued: ROWS_QUEUED.with_label_values(labels),
+            messages_in_flight: MESSAGES_IN_FLIGHT.with_label_values(labels),
+        }
+    }
+}
+
 #[derive(Clone)]
-pub struct SinkProducerContext;
+pub struct SinkProducerContext {
+    metrics: Arc<SinkMetrics>,
```

Comment:
Prometheus metrics are already `Arc`'d internally, so this is actually a double `Arc`. Just `#[derive(Clone)]` on the `SinkMetrics` and you should be able to ditch the `Arc`!

Reply:
done!